moviebot/user_modeling/user_model.py

```diff
@@ -0,0 +1,152 @@
+            movie_choices = json.load(historical_movie_choices_path)
+            self._movies_choices.update(movie_choices)
+
+        self.tag_preferences = defaultdict(defaultdict(float))
+
+    @property
+    def movie_choices(self) -> Dict[str, str]:
+        """Returns user 's movie choices."""
+        return self._movies_choices
+
+    def update_movie_choice(self, movie_id: str, choice: str) -> None:
+        """Updates the choices for a given user.
+
+        Args:
+            movie_id: Id of the movie.
+            choice: User choice (i.e., accept, reject).
+        """
+        self._movies_choices[movie_id] = self._movies_choices[movie_id] + [
+            choice
+        ]
+
+    def update_movies_choices(self, movies_choices: Dict[str, str]) -> None:
+        """Updates the movie choices for a given user.
+
+        Args:
+            movies_choices: Dictionary with movie choices (i.e., accept,
+              reject).
+        """
+        self._movies_choices.update(movies_choices)
+
+    def get_movie_choices(self, movie_id: str) -> List[str]:
+        """Returns the choices for a given movie.
+
+        Args:
+            movie_id: Id of the movie.
+
+        Returns:
+            List of previous choices for a movie.
+        """
+        return self._movies_choices[movie_id]
+
+    def _convert_choice_to_preference(self, choice: str) -> float:
+        """Converts a choice to a preference within the range [-1,1].
+
+        Dislike is represented by a preference below 0, while like is
+        represented by a preference above 0. If the choice does not express a
+        preference (i.e., inquire), then the preference is neutral, i.e., 0.
+        Possible choices are: accept, reject, dont_like, inquire, and watched.
+
+        Args:
+            choice: Choice (i.e., accept, reject).
+
+        Returns:
+            Preference within the range [-1,1].
+        """
+        if choice == "accept":
+            return 1.0
+        elif choice in ["reject", "dont_like"]:
+            return -1.0
+
+        return 0.0
+
+    def compute_tag_preference(
+        self, slot: str, tag: str, database: DataBase
+    ) -> str:
+        """Computes the preference for a given tag (e.g., comedies).
+
+        Args:
+            slot: Slot name.
+            tag: Tag.
+            database: Database with all the movies.
+
+        Returns:
+            Tag preference.
+        """
+        sql_cursor = database.sql_connection.cursor()
+        tag_set = sql_cursor.execute(
+            f"SELECT ID FROM {database._get_table_name()} WHERE {slot} LIKE "
+            f"'%{tag}%'"
+        ).fetchall()
+
+        preference = 0.0
+        count_rated = 0
+        for movie_id, choices in self._movies_choices.items():
+            if movie_id in tag_set:
+                # TODO: decide how to handle contradictory choices (e.g., the
+                # same movie was accepted and rejected)
+                for choice in choices:
```

Comment:
Why do we have multiple choices for each movie? If a user changes their mind, shouldn't we simply use the latest preference?